pkg/storage/mvcc.go

```diff
@@ -2440,6 +2448,176 @@ func MVCCDeleteRange(
+		return true
+	}
+
+	flushDeleteKeys := func(nonMatch MVCCKey) error {
+		if runSize > rangeTombstoneThreshold ||
+			// Even if we didn't get a large enough number of keys to switch to
+			// using range tombstones, the byte size of the keys we did get is now too large to
+			// encode them all within the byte size limit, so use a range tombstone anyway.
+			batchByteSize+runByteSize >= maxBatchByteSize {
+			if err := ExperimentalMVCCDeleteRangeUsingTombstone(ctx, rw, ms,
+				runStart.Key, nonMatch.Key, endTime, localTimestamp, leftPeekBound, rightPeekBound,
+				0); err != nil {
+				return err
+			}
+			batchByteSize += int64(runStart.EncodedSize() + nonMatch.EncodedSize())
+			batchSize++
+		} else if runSize > 0 {
+			// Use Point tombstones
+			batchByteSize += runByteSize
+			batchSize += int64(runSize)
+			_, _, _, err := MVCCDeleteRange(
+				ctx, rw, ms, runStart.Key, nonMatch.Key,
+				0, endTime, localTimestamp, nil, false)
+			if err != nil {
+				return err
+			}
+		}
+		runSize = 0
+		runStart = MVCCKey{}
+		runEnd = MVCCKey{}
+		return nil
+	}
+
+	// Using the IncrementalIterator with the time-bound iter optimization could
+	// potentially be a big win here -- the expected use-case for this is to run
+	// over an entire table's span with a very recent timestamp, rolling back just
+	// the writes of some failed IMPORT and that could very likely only have hit
+	// some small subset of the table's keyspace. However to get the stats right
+	// we need a non-time-bound iter e.g. we need to know if there is an older key
+	// under the one we are clearing to know if we're changing the number of live
+	// keys. The MVCCIncrementalIterator uses a non-time-bound iter as its source
+	// of truth, and only uses the TBI iterator as an optimization when finding
+	// the next KV to iterate over. This pattern allows us to quickly skip over
+	// swaths of uninteresting keys, but then use a normal iteration to actually
+	// do the delete including updating the live key stats correctly.
+	//
+	// The MVCCIncrementalIterator checks for and fails on any intents in our
+	// time-range, as we do not want to clear any running transactions. We don't
+	// _expect_ to hit this since the RevertRange is only intended for non-live
+	// key spans, but there could be an intent leftover.
+	iter := NewMVCCIncrementalIterator(rw, MVCCIncrementalIterOptions{
+		EndKey:    endKey,
+		StartTime: predicates.ExperimentalPredicateTime,
+		EndTime:   endTime,
+	})
+	defer iter.Close()
+
+	iter.SeekGE(MVCCKey{Key: startKey})
+	for {
+		if ok, err := iter.Valid(); err != nil {
+			return nil, err
+		} else if !ok {
+			break
+		}
+
+		k := iter.UnsafeKey()
+		vRaw := iter.UnsafeValue()
+		v, err := DecodeMVCCValue(vRaw)
+		if err != nil {
+			return nil, err
+		}
+		if k.Key.Equal(runEnd.Key) {
+			// If we've already added this key to the run, move on.
+			//
+			// TODO(msbutler): we could introduce iter.NextKeyIgnoringTime(
```

Comment:
Should we code up `iter.NextKeyIgnoringTime()`? Or are there correctness issues that I'm missing? Once we've matched a key, I don't see why we need to surface it's older versions.

Reply:
I'll have to look closer at this, but it's late here -- will get back to you on this tomorrow.

Reply:
Yeah, I think you're right. Kind of annoying that I just removed `NextKeyIgnoringTime` over in #82691. :) Wanna submit a PR for it?